Unpin Firedrake image for CI

pbrubeck opened the pull request:
Solves https://github.com/g-adopt/g-adopt/issues/135

angus-g commented:
Thanks Pablo! A bit annoying (but understandable) that GitHub doesn't pass secrets through to external forks so the CI breaks... I will put that on my list.

I've run the test from within the organisation, and it looks like your changes are fine. We still can't unpin because there's another UFL-related failure. In the Spiegelman cases that do stabilisation and modify the Jacobian, there is the following failure. The same failure comes up in the Schmalholz multi-material benchmark (I guess something about how its Jacobian is constructed?)

```
Traceback (most recent call last):
  File "/__w/g-adopt/g-adopt/tests/Drucker-Prager_rheology/spiegelman.py", line 271, in <module>
    spiegelman(
  File "/__w/g-adopt/g-adopt/tests/Drucker-Prager_rheology/spiegelman.py", line 255, in spiegelman
    newton_solver.solve()
  File "/home/firedrake/firedrake/lib/python3.12/site-packages/gadopt/stokes_integrators.py", line 473, in solve
    self.setup_solver()
  File "/home/firedrake/firedrake/lib/python3.12/site-packages/gadopt/stokes_integrators.py", line 460, in setup_solver
    self.solver = fd.NonlinearVariationalSolver(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/usr/lib/python3.12/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint_utils/variational_solver.py", line 47, in wrapper
    init(self, problem, *args, **kwargs)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/variational_solver.py", line 247, in __init__
    ctx.set_jacobian(self.snes)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/solving_utils.py", line 288, in set_jacobian
    snes.setJacobian(self.form_jacobian, J=self._jac.petscmat,
                                           ^^^^^^^^^
  File "/usr/lib/python3.12/functools.py", line 995, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/solving_utils.py", line 490, in _jac
    return self._assembler_jac.allocate()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1330, in allocate
    self._make_maps_and_regions())
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1359, in _make_maps_and_regions
    elif any(local_kernel.indices == (None, None) for assembler in self._all_assemblers for local_kernel, _ in assembler.local_kernels):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1359, in <genexpr>
    elif any(local_kernel.indices == (None, None) for assembler in self._all_assemblers for local_kernel, _ in assembler.local_kernels):
                                                                                                               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/functools.py", line 995, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1069, in local_kernels
    kernels = tsfc_interface.compile_form(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/firedrake/tsfc_interface.py", line 220, in compile_form
    tsfc_kernel = TSFCKernel(
                  ^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/tsfc_interface.py", line 95, in __init__
    tree = tsfc_compile_form(form, prefix=name, parameters=parameters,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/tsfc/driver.py", line 67, in compile_form
    fd = ufl_utils.compute_form_data(form, complex_mode=complex_mode)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/tsfc/ufl_utils.py", line 56, in compute_form_data
    fd = ufl_compute_form_data(
         ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/ufl/ufl/algorithms/compute_form_data.py", line 423, in compute_form_data
    check_form_arity(preprocessed_form, self.original_form.arguments(), complex_mode)
  File "/home/firedrake/firedrake/src/ufl/ufl/algorithms/check_arities.py", line 213, in check_form_arity
    check_integrand_arity(itg.integrand(), arguments, complex_mode)
  File "/home/firedrake/firedrake/src/ufl/ufl/algorithms/check_arities.py", line 194, in check_integrand_arity
    arg_tuples = map_expr_dag(rules, expr, compress=False)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/ufl/ufl/corealg/map_dag.py", line 35, in map_expr_dag
    (result,) = map_expr_dags(
                ^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/ufl/ufl/corealg/map_dag.py", line 103, in map_expr_dags
    r = handlers[v._ufl_typecode_](v, *[vcache[u] for u in v.ufl_operands])
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/ufl/ufl/algorithms/check_arities.py", line 60, in sum
    f"Adding expressions with non-matching form arguments {_afmt(a)} vs {_afmt(b)}."
                                                                         ^^^^^^^^
  File "/home/firedrake/firedrake/src/ufl/ufl/algorithms/check_arities.py", line 20, in _afmt
    arg, conj = atuple
    ^^^^^^^^^
ValueError: not enough values to unpack (expected 2, got 1)
```

It's a pretty deep and miserable backtrace. The `a` object in the second-last frame is:
````
(
    (Argument(WithGeometry(NoneProxyFunctionSpace(<firedrake.mesh.MeshTopology object at 0x7bff459a2810>, VectorElement(FiniteElement('Q', quadrilateral, 2), dim=2), name=None, index=None, component=None), Mesh(VectorElement(FiniteElement('Q', quadrilateral, 1), dim=2), 0)), 0, None), False),
    (Argument(WithGeometry(NoneProxyFunctionSpace(<firedrake.mesh.MeshTopology object at 0x7bff459a2810>, FiniteElement('Q', quadrilateral, 1), name=None, index=None, component=None), Mesh(VectorElement(FiniteElement('Q', quadrilateral, 1), dim=2), 0)), 1, None), False)
)
```
and `b` is missing the conj part of the tuple, but the arg part is identical (same `id` value):
```
(
    (Argument(WithGeometry(NoneProxyFunctionSpace(<firedrake.mesh.MeshTopology object at 0x7bff459a2810>, VectorElement(FiniteElement('Q', quadrilateral, 2), dim=2), name=None, index=None, component=None), Mesh(VectorElement(FiniteElement('Q', quadrilateral, 1), dim=2), 0)), 0, None), False),
)
```

angus-g commented:
In fact, it looks like this UFL commit is the culprit: https://github.com/firedrakeproject/ufl/commit/cb9052da7cd8b888400734aa248ee9d1cff1a371
All tests run and pass with its parent, but not with it.

pbrubeck commented:
Could you point me to the test?

angus-g commented:
In the `tests/Drucker-Prager_rheology` directory, you can run `python3 spiegelman.py 0.0025_1e+23_64_16_15_True`. There's also `python3 run_benchmark.py benchmark.schmalholz_2011` from the `tests/multi_material` directory.

pbrubeck commented:
I ran `make test`, but I still get a `FileNotFoundError`:
```
    def residual(ui, mui, nx, ny, picard_iterations, stabilisation):
        output_dir = Path(__file__).parent.resolve() / f"spiegelman_{ui}_{mui}_{nx}_{ny}_{picard_iterations}_{stabilisation}"
    
>       with open(output_dir / "picard.txt", "r") as f:
E       FileNotFoundError: [Errno 2] No such file or directory: '/home/pbrubeck/firedrake/src/g-adopt/tests/Drucker-Prager_rheology/spiegelman_0.005_1e+24_64_16_50_False/picard.txt'
```

angus-g commented:
> I ran `make test`, but I still get a `FileNotFoundError`

There's a bit of fiddling to get your environment set up to make that work. You could use `make check` within the directories instead. Your fix in 1a2dfd8 works, but there's still the failure in multi-material. Seems the Schmalholz case uses a sharp viscosity transition https://github.com/g-adopt/g-adopt/blob/03ea2a086b325a61263b66d3415612a1c5a8d289/tests/multi_material/run_benchmark.py#L168-L173 which resolves to a `conditional`: https://github.com/g-adopt/g-adopt/blob/03ea2a086b325a61263b66d3415612a1c5a8d289/gadopt/level_set_tools.py#L370-L375

This has the same traceback as before
<details>
<summary>Traceback</summary>

```
Traceback (most recent call last):
  File "/__w/g-adopt/g-adopt/tests/multi_material/run_benchmark.py", line 231, in <module>
    stokes_solver.solve()
  File "/home/firedrake/firedrake/lib/python3.12/site-packages/gadopt/stokes_integrators.py", line 473, in solve
Traceback (most recent call last):
  File "/__w/g-adopt/g-adopt/tests/multi_material/run_benchmark.py", line 231, in <module>
    stokes_solver.solve()
  File "/home/firedrake/firedrake/lib/python3.12/site-packages/gadopt/stokes_integrators.py", line 473, in solve
Traceback (most recent call last):
  File "/__w/g-adopt/g-adopt/tests/multi_material/run_benchmark.py", line 231, in <module>
    stokes_solver.solve()
  File "/home/firedrake/firedrake/lib/python3.12/site-packages/gadopt/stokes_integrators.py", line 473, in solve
    self.setup_solver()
  File "/home/firedrake/firedrake/lib/python3.12/site-packages/gadopt/stokes_integrators.py", line 460, in setup_solver
    self.solver = fd.NonlinearVariationalSolver(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/usr/lib/python3.12/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint_utils/variational_solver.py", line 47, in wrapper
    init(self, problem, *args, **kwargs)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/variational_solver.py", line 247, in __init__
    ctx.set_jacobian(self.snes)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/solving_utils.py", line 288, in set_jacobian
    snes.setJacobian(self.form_jacobian, J=self._jac.petscmat,
                                           ^^^^^^^^^
  File "/usr/lib/python3.12/functools.py", line 995, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/solving_utils.py", line 490, in _jac
    return self._assembler_jac.allocate()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1330, in allocate
    self._make_maps_and_regions())
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1359, in _make_maps_and_regions
    elif any(local_kernel.indices == (None, None) for assembler in self._all_assemblers for local_kernel, _ in assembler.local_kernels):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1359, in <genexpr>
    elif any(local_kernel.indices == (None, None) for assembler in self._all_assemblers for local_kernel, _ in assembler.local_kernels):
                                                                                                               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/functools.py", line 995, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1069, in local_kernels
    kernels = tsfc_interface.compile_form(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
    self.setup_solver()
  File "/home/firedrake/firedrake/lib/python3.12/site-packages/gadopt/stokes_integrators.py", line 460, in setup_solver
    self.solver = fd.NonlinearVariationalSolver(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/usr/lib/python3.12/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint_utils/variational_solver.py", line 47, in wrapper
    init(self, problem, *args, **kwargs)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/variational_solver.py", line 247, in __init__
    ctx.set_jacobian(self.snes)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/solving_utils.py", line 288, in set_jacobian
    snes.setJacobian(self.form_jacobian, J=self._jac.petscmat,
                                           ^^^^^^^^^
  File "/usr/lib/python3.12/functools.py", line 995, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/solving_utils.py", line 490, in _jac
    return self._assembler_jac.allocate()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1330, in allocate
    self._make_maps_and_regions())
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1359, in _make_maps_and_regions
    elif any(local_kernel.indices == (None, None) for assembler in self._all_assemblers for local_kernel, _ in assembler.local_kernels):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1359, in <genexpr>
    elif any(local_kernel.indices == (None, None) for assembler in self._all_assemblers for local_kernel, _ in assembler.local_kernels):
                                                                                                               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/functools.py", line 995, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1069, in local_kernels
    kernels = tsfc_interface.compile_form(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/firedrake/tsfc_interface.py", line 220, in compile_form
    tsfc_kernel = TSFCKernel(
                  ^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/tsfc_interface.py", line 95, in __init__
    tree = tsfc_compile_form(form, prefix=name, parameters=parameters,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
Traceback (most recent call last):
  File "/__w/g-adopt/g-adopt/tests/multi_material/run_benchmark.py", line 231, in <module>
    stokes_solver.solve()
  File "/home/firedrake/firedrake/lib/python3.12/site-packages/gadopt/stokes_integrators.py", line 473, in solve
    self.setup_solver()
  File "/home/firedrake/firedrake/lib/python3.12/site-packages/gadopt/stokes_integrators.py", line 460, in setup_solver
    self.solver = fd.NonlinearVariationalSolver(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/usr/lib/python3.12/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint_utils/variational_solver.py", line 47, in wrapper
    init(self, problem, *args, **kwargs)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/variational_solver.py", line 247, in __init__
    ctx.set_jacobian(self.snes)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/solving_utils.py", line 288, in set_jacobian
    snes.setJacobian(self.form_jacobian, J=self._jac.petscmat,
                                           ^^^^^^^^^
  File "/usr/lib/python3.12/functools.py", line 995, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/solving_utils.py", line 490, in _jac
    return self._assembler_jac.allocate()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1330, in allocate
    self._make_maps_and_regions())
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1359, in _make_maps_and_regions
    elif any(local_kernel.indices == (None, None) for assembler in self._all_assemblers for local_kernel, _ in assembler.local_kernels):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1359, in <genexpr>
    elif any(local_kernel.indices == (None, None) for assembler in self._all_assemblers for local_kernel, _ in assembler.local_kernels):
                                                                                                               ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/functools.py", line 995, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1069, in local_kernels
    kernels = tsfc_interface.compile_form(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
    self.setup_solver()
  File "/home/firedrake/firedrake/lib/python3.12/site-packages/gadopt/stokes_integrators.py", line 460, in setup_solver
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
    self.solver = fd.NonlinearVariationalSolver(
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/firedrake/tsfc_interface.py", line 220, in compile_form
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/usr/lib/python3.12/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/adjoint_utils/variational_solver.py", line 47, in wrapper
    init(self, problem, *args, **kwargs)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/variational_solver.py", line 247, in __init__
    ctx.set_jacobian(self.snes)
  File "/home/firedrake/firedrake/src/firedrake/firedrake/solving_utils.py", line 288, in set_jacobian
    snes.setJacobian(self.form_jacobian, J=self._jac.petscmat,
                                           ^^^^^^^^^
  File "/usr/lib/python3.12/functools.py", line 995, in __get__
    tsfc_kernel = TSFCKernel(
                  ^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/tsfc_interface.py", line 95, in __init__
    tree = tsfc_compile_form(form, prefix=name, parameters=parameters,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/tsfc/driver.py", line 67, in compile_form
    fd = ufl_utils.compute_form_data(form, complex_mode=complex_mode)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/tsfc/ufl_utils.py", line 56, in compute_form_data
    fd = ufl_compute_form_data(
         ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/ufl/ufl/algorithms/compute_form_data.py", line 423, in compute_form_data
    check_form_arity(preprocessed_form, self.original_form.arguments(), complex_mode)
  File "/home/firedrake/firedrake/src/ufl/ufl/algorithms/check_arities.py", line 213, in check_form_arity
    check_integrand_arity(itg.integrand(), arguments, complex_mode)
  File "/home/firedrake/firedrake/src/ufl/ufl/algorithms/check_arities.py", line 194, in check_integrand_arity
    arg_tuples = map_expr_dag(rules, expr, compress=False)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/ufl/ufl/corealg/map_dag.py", line 35, in map_expr_dag
    (result,) = map_expr_dags(
                ^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/ufl/ufl/corealg/map_dag.py", line 103, in map_expr_dags
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
    value = func(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/firedrake/tsfc_interface.py", line 220, in compile_form
    tsfc_kernel = TSFCKernel(
                  ^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/tsfc_interface.py", line 95, in __init__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/solving_utils.py", line 490, in _jac
            ^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/tsfc/driver.py", line 67, in compile_form
    r = handlers[v._ufl_typecode_](v, *[vcache[u] for u in v.ufl_operands])
    tree = tsfc_compile_form(form, prefix=name, parameters=parameters,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "petsc4py/PETSc/Log.pyx", line 188, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
    return self._assembler_jac.allocate()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/firedrake/assemble.py", line 1330, in allocate
    fd = ufl_utils.compute_form_data(form, complex_mode=complex_mode)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/firedrake/tsfc/ufl_utils.py", line 56, in compute_form_data
    fd = ufl_compute_form_data(
         ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/ufl/ufl/algorithms/compute_form_data.py", line 423, in compute_form_data
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/firedrake/firedrake/src/ufl/ufl/algorithms/check_arities.py", line 60, in sum
    f"Adding expressions with non-matching form arguments {_afmt(a)} vs {_afmt(b)}."
                                                                         ^^^^^^^^
  File "/home/firedrake/firedrake/src/ufl/ufl/algorithms/check_arities.py", line 20, in _afmt
    arg, conj = atuple
    ^^^^^^^^^
ValueError: not enough values to unpack (expected 2, got 1)
application called MPI_Abort(PYOP2_COMM_WORLD, 1) - process 4
  File "petsc4py/PETSc/Log.pyx", line 189, in petsc4py.PETSc.Log.EventDecorator.decorator.wrapped_func
  File "/home/firedrake/firedrake/src/firedrake/pyop2/caching.py", line 550, in wrapper
```
</details>

But as far as I can tell, the Jacobian isn't being manually set in this case. As with before, the failure was introduced by https://github.com/firedrakeproject/ufl/commit/cb9052da7cd8b888400734aa248ee9d1cff1a371, and the case works as-is if this commit is reverted.

pbrubeck commented:
How do I reproduce the multimaterial failing test? Which arguments do I need to pass `run_benchmark.py`? 

It'd be a lot easier if the tests are set up to run with pytest...

angus-g commented:
> How do I reproduce the multimaterial failing test? Which arguments do I need to pass `run_benchmark.py`?

As I said [above](https://github.com/g-adopt/g-adopt/pull/156#issuecomment-2600150565), you can run that specific case with
```
python3 run_benchmark.py benchmarks.schmalholz_2011
```

You can run all the tests for a given directory with `make check`, but you may have to run `make clean` first to clean previous output files. For the multi-material tests, there's like 9 of them that use 8 cores each, so it can take a while. Given the failure is just when *running* the test, I would just run it specifically...

angus-g commented:
More specifically, it's [this part](https://github.com/FEniCS/ufl/commit/0a0a4dfff4af651b2099d01834244126f1cb103a#diff-dea4762171ede7bfe8a50e06ee8255cd21ea0fc34c34760591971c21244aff88R29-R36) of the `ListTensor` simplification that causes the failure.

An `Indexed` is created with the indices `MultiIndex((FixedIndex(1), Index(29))`, which doesn't really seem to fit any of the optimisation patterns outlined in https://github.com/FEniCS/ufl/pull/336, so it drops part of the expression in the derivative, which I assume is what causes the error during the Jacobian construction.

angus-g commented:
Looks like there's also about a 30% performance regression on upstream vs. the last working version. Same machine, both run today: [current](https://github.com/g-adopt/g-adopt/actions/runs/12878906036/attempts/1#summary-35905549311) vs. [proposed](https://github.com/g-adopt/g-adopt/actions/runs/12877703273/attempts/1#summary-35902566120)

For example, the mantle convection base case jumps from 70s runtime to 172s. It looks like the reference velocity computation causes kernel re-compilation every time:
https://github.com/g-adopt/g-adopt/blob/03ea2a086b325a61263b66d3415612a1c5a8d289/gadopt/utility.py#L82

The timings between the two are broadly similar, but we can see the timestep computation is responsible for most of the change in runtime:
![image](https://github.com/user-attachments/assets/a1520850-025f-4d9d-9c41-270732f250a9)

pbrubeck commented:
> An `Indexed` is created with the indices `MultiIndex((FixedIndex(1), Index(29))`, which doesn't really seem to fit any of the optimisation patterns outlined in [FEniCS/ufl#336](https://github.com/FEniCS/ufl/pull/336), so it drops part of the expression in the derivative, which I assume is what causes the error during the Jacobian construction.

That's the mechanism that simplifies [v0, ..., vk][j] -> vj, and the constructor `Indexed()` does not necessarily return an `Indexed` instance, it first tries to select the correct element or slice if possible.

The issue is that the simplications that we introduced were spuriously creating a `conditional(cond, 0, 0)` that was added to a `TestFunction` for forms with `derivative(conditional(expr(split(z)[k])), z)`. We have addressed this in https://github.com/FEniCS/ufl/pull/340

Both test failures should be fixed once https://github.com/firedrakeproject/ufl/pull/58 is merged, and no changes would be required on these tests.

angus-g commented:
Excellent, looks like this will be good to go pending those few PRs. Thanks @pbrubeck!

angus-g reviewed (approved):
Passes tests: https://github.com/g-adopt/g-adopt/actions/runs/12916727045

Just unpinned the image for rendering docs too. Good to go, thanks again for your help on this!
